### PR TITLE
drivers: add support for the Bosch bmx055 9-axis IMU

### PR DESCRIPTION
**Pull request by SemjonKerner**
This is the initial driver implementation for the bmx055 9-axis sensor. This implementation provides basic functions for initialization and reading sensor data from all three modules (magnetometer, accelerometer, gyroscope) and also adds support for saul. Since the module is part of the calliope-mini, it is also added to the boards dependencies.

### CODE REVIEW

File: drivers/bmx055/bmx055.c [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+ *
+ * @return                  BMX055_OK on success
+ * @return                  BMX055_NOBUS if i2C connection can not be establish
+ * @return                  BMX055_NODEV if a module can not be read out
+ * @return                  BMX055_NOWRITE if a register can not be written
+ * @return                  BMX055_NOREAD if a register can not be read
+ */
+int bmx055_init(bmx055_t *dev, const bmx055_params_t *params)
+{
+	assert(dev && params);
```
Comment (aabadie) [inline drivers/bmx055/bmx055.c]: indentation is off in this file (are you mixing tabs and spaces ?)

File: drivers/bmx055/bmx055.c [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
+#define BUS_CLK  		(I2C_SPEED_FAST)
+#define BUS  			(dev->p.i2c)
+#define ADDR_MAG		(dev->p.addr_mag)
+#define ADDR_ACC		(dev->p.addr_acc)
+#define ADDR_GYRO		(dev->p.addr_gyro)
+#define GYRO_2000_DPS   (2000U)
+#define GYRO_MAX_SCALE  (0x7FFFU)
+
+/**
```
Comment (aabadie) [inline drivers/bmx055/bmx055.c]: The doxygen block is a duplicate from the header one. Same for all functions in this file

File: drivers/bmx055/bmx055_internal.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * @author      Semjon Kerner <semjon.kerner@fu-berlin.de>
+ */
+
+#ifndef BMX055_INTERNAL_H
+#define BMX055_INTERNAL_H
+
+/* MAGNETOMETER */
+
+#define REG_MAG_CHIPID      (0x40U)
```
Comment (aabadie) [inline drivers/bmx055/bmx055_internal.h]: Please add some doxygen documentation in this file (only `@name` blocks are required I think)

File: drivers/bmx055/bmx055.c [hunk 4]
```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (C) 2017 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_bmx055 BMX055 9-axis gyroscope
```
Comment (aabadie) [inline drivers/bmx055/bmx055.c]: Copy-pasta from header file. Please adapt

File: drivers/include/bmx055.h [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   The sensors default I2C addresses
+ */
+#define BMX055_MAG_ADDR_DEFAULT  0x10
+#define BMX055_ACC_ADDR_DEFAULT  0x18
```
Comment (aabadie) [inline drivers/include/bmx055.h]: Those 2 defines are not documented. Use `@name` around the 3 or a `@brief` for each ones

File: tests/driver_bmx055/README.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# About
+This is a test application that is polling all three SAUL devices
+(Magnetometer, Accelerometer, Gyroscope) of the Bosch BMX055
+9-Axis Sensor.
+This test is a copy of the standard saul test. Since it utilizes all
```
Comment (aabadie) [inline tests/driver_bmx055/README.md]: I would prefer a minimal test application that doesn't use the saul module but just the driver API. This is how things are done for other drivers and we want to be consistent.

File: drivers/include/bmx055.h [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+extern "C" {
+#endif
+
+/**
+ * @name   The sensors default I2C addresses
+ * @{
+ */
+#define BMX055_MAG_ADDR_DEFAULT  0x10
+#define BMX055_ACC_ADDR_DEFAULT  0x18
+#define BMX055_GYRO_ADDR_DEFAULT 0x68
```
Comment (haukepetersen) [inline drivers/include/bmx055.h]: minor style: its good practice to put parenthesis around scalar values in defines

File: sys/auto_init/saul/auto_init_bmx055.c [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        int res = bmx055_init(&bmx055_devs[i], &bmx055_params[i]);
+        if (res != BMX055_OK) {
+            LOG_ERROR("[auto_init_saul] error initializing bmx055 #%u\n", i);
+            continue;
+        }
+
+        /*
+         * magnetometer
+         */
+        saul_entries[0].dev = &(bmx055_devs[i]);
```
Comment (haukepetersen) [inline sys/auto_init/saul/auto_init_bmx055.c]: the array offsets in these lines are broken. Say you have 2 BMX055 devices, then the initialization of the 2nd device will overwrite the registry entry for the first device... So using `saul_entries[((i * 3) + 0)]. ...`, and `+1` and `+2` should do the trick

File: tests/driver_bmx055/README.md (hunk 6 above)
Comment (SemjonKerner) [reply tests/driver_bmx055/README.md]: I looked over the example you provided, but i also talked to @haukepetersen about it and he recommended using the saul test. I am not in the position to decide, so I would prefer both of you decide this together. I would be okay writing another test.

File: drivers/bmx055/bmx055.c [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+    res = i2c_read_reg(BUS, ADDR_ACC, REG_ACC_CHIPID, &tmp);
+    if (res != 1 || tmp != REG_ACC_CHIPID_VAL) {
+        DEBUG("[bmx055] error: no connection to accelerometer\n");
+        return BMX055_NODEV;
+    }
+
+    /*
+     * init range parameter data
+     */
+    switch (dev->p.acc_range) {
```
[the remark this answers is not in the document]
Comment (haukepetersen) [reply drivers/bmx055/bmx055.c]: Or alternatively, if you use the enum values are also used, you could do something like
```c
// on top of this file:
static const uint8_t acc_ranges[] = {
    REG_ACC_RANGE_2G,
    REG_ACC_RANGE_4G,
    ...

};

// and in this function:
range = acc_ranges[dev->p.acc_range];
```

File: tests/driver_bmx055/README.md (hunk 6 above)
Comment (smlng) [reply tests/driver_bmx055/README.md]: I agree with @aabadie to have a non-saul test application using the drivers _API_, for testing with SAUL there already is `tests/saul` and also `examples/saul` plus `examples/default` would also do. I mean, in the worst case this tests fails due to some issue with SAUL and not with the driver itself.

**Comment by haukepetersen** [comment]
@aabadie, @smlng: the supplied test application is sufficient and the way to go here. It provides all things necessary for testing the `bmx055` driver:
- enables access to all code paths of the driver
- also tests the SAUL adaption layer
- forces to the CI to build the driver (-> `USEMODULE += bmx055` in the tests Makefile)

I agree that simply copying the `test/saul` application with this one-line change in the Makefile is by far not optimal, but as long as our (CI-)test system does not support building test applications parameterized, this is I guess a pill we have to swallow. At some point in the future, I actually expect almost all driver specific tests to be deleted in favor of something like `build test/saul for boards [x, y, z, ...] with one USEMODULE out of [bmx055, aa, bb, cc, ...]`.

**Comment by aabadie** [comment]
I'm sorry but I tend to disagree with you @haukepetersen: I see test applications as good examples for using the drivers API. If one day, we decide to only have one that uses saul, we'll loose some kind of useful documentation for newcomers.

**Comment by haukepetersen** [comment]
> I'm sorry but I tend to disagree with you @haukepetersen: I see test applications as good examples for using the drivers API. If one day, we decide to only have one that uses SAUL, we'll loose some kind of useful documentation for newcomers.

No, really? I don't see where the test is useful documentation! For this particular driver (and also most others), the driver specific API is trivial, and having a (trivial) test case using these functions directly is not really adding anything to the table then to waste time for the person forced to implement it. Also, one can always look into the SAUL adaption code of the driver to see how its functions can be used...

**Comment by SemjonKerner** [comment]
I can see there are different views, that you should discuss properly and compromise somewhen in the near future. Until then I would suggest leaving the saul test in this PR as is, especially because it is a proper test for the driver, as hauke explained.

File: drivers/bmx055/bmx055.c [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    }
+    if (tmp != REG_GYRO_CHIPID_VAL) {
+        DEBUG("[bmx055] error: no connection to gyroscope\n");
+        return BMX055_NODEV;
+    }
+
+    /* Init Magnetometer
+     *
+     * set magnetometer to normal mode (Bits 1 & 2 = 0x00)
+     * and set magnetometer sample rate (Bits 3 to 5) */
```
Comment (smlng) [inline drivers/bmx055/bmx055.c]: minor nit, place `*/` on new line, i.e.

```
 * and set magnetometer sample rate (Bits 3 to 5) 
 */
```

apply below, too

Comment (smlng) [reply drivers/bmx055/bmx055.c]: for multi line comments only, keep one-liner as is 😉

**Comment by smlng** [comment]
>  seems that we disagree on the test. So do you insist?

For me it would be fine to go with the SAUL test (for now). But in general I would prefer distinct driver tests, as SAUL is (kind of) limited in its functionality, it might not make use of every call/feature a driver provides. Further, if there is a problem in SAUL every such driver test would fail though there is no issue with driver itself.

TL;DR: no I don't insist, right now.

File: drivers/bmx055/bmx055.c (hunk 10 above)
Comment (SemjonKerner) [reply drivers/bmx055/bmx055.c]: Okay, should I also add a new line infront of a multi line comment then, like
```
/*
 * Init Magnetometer
 * ...
 */
```

**Comment by haukepetersen** [comment]
> as SAUL is (kind of) limited in its functionality, it might not make use of every call/feature a driver provides. 

I do agree with this in general, but in the case of this driver (and many others as well), the SAUL mapping does actually execute/trigger all possible functions/code paths of the driver and is hence complete, so from a coverage point of view no necessity for a more advanced test application. Different reasoning for devices that expose functionality that exceeds the one used by the SAUL mapping, of course...

> Further, if there is a problem in SAUL every such driver test would fail though there is no issue with driver itself.

True, but such a fail is hopefully noticed and fixed quite fast (as it effects to many drivers), so that I'd say that it is safe to assume that SAUL (and the SAUL registry) work as expected, right?!

> TL;DR: no I don't insist, right now.

Thanks :-)

**Comment by aabadie** [comment]
Even if I understand all the technical considerations and limitations exposed above, I'm still not convinced from a documentation point of view.
With the SAUL approach, if I'm new to RIOT and just want to use this sensor, I'll have to understand the concepts of SAUL and if I really don't want to use SAUL, I'll have to dig in the driver implementation to get an example usage of the API. And it wouldn't even directly show how to initialize the sensor (this is done in sys/auto_init) but just how to read/configure it.
On the contrary, if just the driver API is used in the test application (like in most other cases), I can simply start from there and it saves me a lot of time.

So what I would propose (let's be constructive, otherwise this might never end):
- either drop SAUL and use the driver API in the test application
- either keep saul, but extends the driver doxygen documentation with a small snippet showing the usage of the driver API

File: drivers/bmx055/bmx055.c [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+        DEBUG("[bmx055] error: setting accelerometer range\n");
+        return BMX055_NOWRITE;
+    }
+
+    /* enable acc shadowing */
+    if (i2c_write_reg(BUS, ADDR_ACC, REG_ACC_SHDW, REG_ACC_SHDW_ENABLE) != 1) {
+        DEBUG("[bmx055] error: writing accelerometer shadowing bit\n");
+        return BMX055_NOWRITE;
+    }
+
```
Comment (kaspar030) [inline drivers/bmx055/bmx055.c]: remove newline

File: drivers/include/bmx055.h [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    uint8_t gyro_scale;                 /**< range of gyroscope*/
+} bmx055_params_t;
+
+/**
+ * @brief   Device descriptor for BMX055 sensors
+ */
+typedef struct {
+    bmx055_params_t p;  /**< Device initialization parameters*/
+} bmx055_t;
+
```
Comment (kaspar030) [inline drivers/include/bmx055.h]: remove extra newline

File: sys/auto_init/saul/auto_init_bmx055.c [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+ */
+static saul_reg_t saul_entries[BMX055_NUM * BMX055_SENSOR_NUM];
+
+/**
+ * @brief   Reference the driver struct
+ */
+extern saul_driver_t bmx055_magnetometer_saul_driver;
+extern saul_driver_t bmx055_accelerometer_saul_driver;
+extern saul_driver_t bmx055_gyroscope_saul_driver;
+
```
Comment (kaspar030) [inline sys/auto_init/saul/auto_init_bmx055.c]: remove extra newline

**Comment by miri64** [comment]
Honestly, if our documentation is so bad that we need *tests* to explain a component, we should rethink our documentation approach, not have a test explaining the very simple API... I think that both the documentation for this driver is sufficient enough for newcomers to be understood and that the tests show that the driver is functioning (there is even a README that explains how to use it). Please don't block the PR of a new contributor because it isn't in line with your personal style.

**Comment by kaspar030** [comment]
> I see test applications as good examples for using the drivers API.

They might often be, but they don't have to be. "tests/" is foremost for testing. I'm fine with a saul-based test. The worst that can happen is that a new user is confused and thinks he has to use the driver through saul...

**Comment by aabadie** [comment]
>  it isn't in line with your personal style

This is not **my** personal style but rather a general rule that we should always follow. I know the API is very simple but that's not a good reason. Good documentation is as important as code good code/APIs for an open-source project.

Let's stop *blocking* this PR and move forward here and in other places.

**Comment by kaspar030** [comment]
> This is not my personal style but rather a general rule that we should always follow.

:)

**Comment by miri64** [comment]
> This is not my personal style but rather a general rule that we should always follow.

If a rule hinders progress without providing any real benefit (which IMHO this rule I never heard of does), it should be re-thought ;-).

**Comment by kaspar030** [comment]
Anyhow, thanks @SemjonKerner for your patience and thanks @aabadie for trying to improve RIOT's documentation. Let's take your proposal to the next f2f!

File: drivers/bmx055/bmx055_internal.h [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+ * @brief       Definitions for the bmx055 device
+ *
+ * @author      Semjon Kerner <semjon.kerner@fu-berlin.de>
+ */
+
+#ifndef BMX055_INTERNAL_H
+#define BMX055_INTERNAL_H
+
+#ifdef __cplusplus
+ extern "C" {
```
Comment (aabadie) [inline drivers/bmx055/bmx055_internal.h]: extra space before `extern`

File: drivers/bmx055/bmx055_saul.c [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * @ingroup     drivers_bmx055
+ * @{
+ *
+ * @file
+ * @brief       bmx055 adaption to the RIOT actuator/sensor interface
+ *
+ * @author      Semjon Kerner <semjon.kerner@fu-berlin.de>
+ *
```
Comment (aabadie) [inline drivers/bmx055/bmx055_saul.c]: This line could be removed

File: sys/auto_init/saul/auto_init_bmx055.c [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+/*
+ * @ingroup     auto_init_saul
+ * @{
+ *
+ * @file
+ * @brief       Auto initialization of bmx055 9-axis sensors
+ *
+ * @author      Semjon Kerner <semjon.kerner@fu-berlin.de>
+ *
```
Comment (aabadie) [inline sys/auto_init/saul/auto_init_bmx055.c]: This line could be removed

File: tests/driver_bmx055/main.c [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * @ingroup     tests
+ *
+ * @file
+ * @brief       Test application for the Bosch BMX055 9-axis Sensor driver
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Semjon Kerner <semjon.kerner@fu-berlin.de>
+ *
```
Comment (aabadie) [inline tests/driver_bmx055/main.c]: This line should be removed

**Comment by SemjonKerner** [comment]
Hello @smlng and @aabadie , can you please recheck your status on this PR, since the requested changes are made.

**Comment by kaspar030** [comment]
Tested OK. ACK. Please squash!

**Comment by haukepetersen** [comment]
@SemjonKerner congrats to your first PR!

**Comment by miri64** [comment]
Congratz @SemjonKerner :tada: :sparkles:
